pkg/install/deploystorage.go

```diff
@@ -285,6 +219,116 @@ func (i *Installer) deployStorageTemplate(ctx context.Context, installConfig *in
 	return i.saveGraph(ctx, g)
 }
 
+func (i *Installer) deploySnapshotUpgradeTemplate(ctx context.Context) error {
+	if os.Getenv("RP_MODE") != "" {
+		// only need this upgrade in production, where there are DenyAssignments
+		return nil
+	}
+
+	resourceGroup := stringutils.LastTokenByte(i.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID, '/')
+
+	clusterSPObjectID, err := i.clusterSPObjectID(ctx)
+	if err != nil {
+		return err
+	}
+
+	t := &arm.Template{
+		Schema:         "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+		ContentVersion: "1.0.0.0",
+		Resources:      []*arm.Resource{i.denyAssignments(clusterSPObjectID)},
+	}
+
+	return i.deployARMTemplate(ctx, resourceGroup, "storage", t, nil)
+}
+
+func (i *Installer) denyAssignments(clusterSPObjectID string) *arm.Resource {
+	notActions := []string{
+		"Microsoft.Network/networkSecurityGroups/join/action",
+	}
+
+	if feature.IsRegisteredForFeature(i.subscriptionDoc.Subscription.Properties, "Microsoft.RedHatOpenShift/EnableSnapshots") {
+		notActions = append(notActions, []string{
+			"Microsoft.Compute/disks/beginGetAccess/action",
+			"Microsoft.Compute/disks/endGetAccess/action",
+			"Microsoft.Compute/disks/write",
+			"Microsoft.Compute/snapshots/beginGetAccess/action",
+			"Microsoft.Compute/snapshots/endGetAccess/action",
+			"Microsoft.Compute/snapshots/write",
+			"Microsoft.Compute/snapshots/delete",
+		}...)
+	}
+
+	return &arm.Resource{
+		Resource: &mgmtauthorization.DenyAssignment{
+			Name: to.StringPtr("[guid(resourceGroup().id, 'ARO cluster resource group deny assignment')]"),
+			Type: to.StringPtr("Microsoft.Authorization/denyAssignments"),
+			DenyAssignmentProperties: &mgmtauthorization.DenyAssignmentProperties{
+				DenyAssignmentName: to.StringPtr("[guid(resourceGroup().id, 'ARO cluster resource group deny assignment')]"),
+				Permissions: &[]mgmtauthorization.DenyAssignmentPermission{
+					{
+						Actions: &[]string{
+							"*/action",
+							"*/delete",
+							"*/write",
+						},
+						NotActions: &notActions,
+					},
+				},
+				Scope: &i.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID,
+				Principals: &[]mgmtauthorization.Principal{
+					{
+						ID:   to.StringPtr("00000000-0000-0000-0000-000000000000"),
+						Type: to.StringPtr("SystemDefined"),
+					},
+				},
+				ExcludePrincipals: &[]mgmtauthorization.Principal{
+					{
+						ID:   &clusterSPObjectID,
+						Type: to.StringPtr("ServicePrincipal"),
+					},
+				},
+				IsSystemProtected: to.BoolPtr(true),
+			},
+		},
+		APIVersion: azureclient.APIVersions["Microsoft.Authorization/denyAssignments"],
+	}
+}
+
+func (i *Installer) clusterSPObjectID(ctx context.Context) (string, error) {
```

Comment:
In the interest of making the diff a bit clearer and also defining this function before it's used, would it make sense to move this before the deployStorageTemplate function?

Reply:
Sure, I can do that